txpool/txpool_test.go

```diff
@@ -121,3 +122,100 @@ func TestTxnQueue_Promotion(t *testing.T) {
 	assert.Equal(t, nonce, uint64(2))
 	assert.Equal(t, pool.Length(), uint64(2))
 }
+
+func TestTxnQueue_Heap(t *testing.T) {
+	type TestCase struct {
+		From     types.Address
+		GasPrice *big.Int
+		Nonce    uint64
+		Index    int
+	}
+
+	addr1 := types.Address{0x1}
+	addr2 := types.Address{0x2}
+
+	test := func(t *testing.T, testTable []TestCase) {
+		pool, err := NewTxPool(hclog.NewNullLogger(), false, &mockStore{}, nil, nil)
+		assert.NoError(t, err)
+		pool.EnableDev()
+
+		for _, testCase := range testTable {
+			err := pool.addImpl("", &types.Transaction{
+				From:     testCase.From,
+				GasPrice: testCase.GasPrice,
+				Nonce:    testCase.Nonce,
+			})
+			assert.NoError(t, err)
+		}
+
+		for _, testCase := range testTable {
+			transaction, _ := pool.Pop()
+
+			assert.NotNil(t, transaction)
+
+			actual := TestCase{
+				From:     transaction.From,
+				GasPrice: transaction.GasPrice,
+				Nonce:    transaction.Nonce,
+			}
+
+			assert.EqualValues(t, testCase, actual)
+		}
+
+		empty, _ := pool.Pop()
+		assert.Nil(t, empty)
+	}
+
+	t.Run("the higher priced transaction should be popped first", func(t *testing.T) {
+		test(t, []TestCase{
+			{
+				From:     addr1,
+				GasPrice: big.NewInt(2),
+			},
+			{
+				From:     addr2,
+				GasPrice: big.NewInt(1),
+			},
+		})
+
+	})
+
+	t.Run("sort by nonce when same from address", func(t *testing.T) {
+		test(t, []TestCase{
+			{
+				From:     addr1,
+				GasPrice: big.NewInt(2),
+				Nonce:    0,
+			},
+			{
+				From:     addr1,
+				GasPrice: big.NewInt(1),
+				Nonce:    1,
+			},
```

Comment:
What do you think of changing the GasPrice so that it is unambiguous that here we actually look at `Nonce` instead of the `GasPrice`?
You could do that by making a second transaction have GasPrice of 3, for example.

Reply:
If I understood correctly, you propose we change the values for the GasPrice so the point of this test case (testing the nonce) is in full focus?

I'll change up the values for the second gas price as you suggested 👍